src/core/lib/event_engine/thread_pool/work_stealing_thread_pool.cc

```diff
@@ -425,36 +456,57 @@ void WorkStealingThreadPool::ThreadState::FinishDraining() {
 // -------- WorkStealingThreadPool::ThreadCount --------
 
 void WorkStealingThreadPool::ThreadCount::Add(CounterType counter_type) {
-  thread_counts_[counter_type].fetch_add(1, std::memory_order_relaxed);
+  grpc_core::MutexLock lock(&wait_mu_[counter_type]);
+  ++thread_counts_[counter_type];
+  wait_cv_[counter_type].SignalAll();
 }
 
 void WorkStealingThreadPool::ThreadCount::Remove(CounterType counter_type) {
-  thread_counts_[counter_type].fetch_sub(1, std::memory_order_relaxed);
+  grpc_core::MutexLock lock(&wait_mu_[counter_type]);
+  --thread_counts_[counter_type];
+  wait_cv_[counter_type].SignalAll();
 }
 
 void WorkStealingThreadPool::ThreadCount::BlockUntilThreadCount(
-    CounterType counter_type, int desired_threads, const char* why,
+    CounterType counter_type, size_t desired_threads, const char* why,
     WorkSignal* work_signal) {
-  auto& counter = thread_counts_[counter_type];
-  int curr_threads = counter.load(std::memory_order_relaxed);
   // Wait for all threads to exit.
-  auto last_log_time = grpc_core::Timestamp::Now();
-  while (curr_threads > desired_threads) {
-    absl::SleepFor(kSleepBetweenQuiesceCheck);
+  while (true) {
+    auto curr_threads = WaitForCountChange(
+        counter_type, desired_threads,
+        grpc_core::Duration::Seconds(kBlockingQuiesceLogRateSeconds));
+    if (curr_threads == desired_threads) break;
+    GRPC_LOG_EVERY_N_SEC_DELAYED(
+        kBlockingQuiesceLogRateSeconds, GPR_DEBUG,
+        "Waiting for thread pool to idle before %s. (%" PRIdPTR " to %" PRIdPTR
+        ")",
+        why, curr_threads, desired_threads);
     work_signal->SignalAll();
```

Comment:
Should there be a work_signal->SignalAll even if curr_threads == desired_threads in the new code ?

Reply:
Good question. `BlockUntilThreadCount` is currently either waiting for:

* 0 threads, in which case nothing is listening on the work_signal when the desired thread count is reached, or
* 1 thread, because this thread is running `BlockUntilThreadCount`, and not waiting on the work_signal either.